src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (size == 0) {
+            head = tail = node;
+        } else {
+            tail.next = node;
+            tail = node;
+
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> node = new Node<>(null, value, null);
+        Node<T> firstNode = head;
+        if (size == 0 || index == size) {
+            add(value);
+        } else if (index == 0) {
+            node = new Node<>(null, value, head);
+            node.next = head;
+            head = node;
+            size++;
+        } else {
+            for (int i = 0; i < index - 1; i++) {
+                firstNode = firstNode.next;
+            }
+            node.next = firstNode.next;
+            node.prev = firstNode;
+            firstNode.next.prev = node;
+            firstNode.next = node;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        T prevValue = getNodeByIndex(index).value;
+        getNodeByIndex(index).value = value;
+
+        return prevValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> node = getNodeByIndex(index);
+        unlink(getNodeByIndex(index));
```

Comment:
Do not call  getNodeByIndex() twice - it has a loop inside and it is two times less productive.